File: lambda/javascript/play-podcast.js
import { GetObjectCommand, waitUntilObjectExists } from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import {
  badRequestErrorResponse,
  getS3Client,
  notFoundErrorResponse,
  PODCAST_BUCKET,
  PODCAST_DATA_KEY_PREFIX,
  temporaryRedirectResponse,
} from './utils.js';

const EPISODE_CONTENT_KIND_MEDIA = 'media';
const EPISODE_CONTENT_KIND_TEXT = 'text';
const HOUR = 60 * 60;

export const handler = async (event) => {
  console.log(`Event:\n${JSON.stringify(event)}`);

  const episodeId = event?.pathParameters.id;
  if (episodeId === undefined) {
    return badRequestErrorResponse('Episode id not provided');
  }

  const contentType = parseEpisodeContentKind(
    event?.queryStringParameters?.content
  );
  if (contentType === undefined) {
    return badRequestErrorResponse(
      `Unknown content kind ${event?.queryStringParameters?.content}`
    );
  }

  const mediaKey = getEpisodeMediaKey(contentType, episodeId);
  if (mediaKey === undefined) {
    return badRequestErrorResponse(`unknonw content type ${contentType}`);
  }

  const s3Client = getS3Client();
  const mediaExists = await checkMediaExists(
    s3Client,
    PODCAST_BUCKET,
    mediaKey
  );
  if (!mediaExists) {
    return notFoundErrorResponse(
      `Episode ${episodeId} ${contentType} data not found`
    );
  }

  return await responseRedirect(s3Client, PODCAST_BUCKET, mediaKey);
};

const parseEpisodeContentKind = (content) => {
  if ([EPISODE_CONTENT_KIND_MEDIA, '', content].includes(content)) {
    return EPISODE_CONTENT_KIND_MEDIA;
  } else if (content === 'text') {
    return EPISODE_CONTENT_KIND_TEXT;
  }
};

const getEpisodeMediaKey = (contentType, episodeId) => {
  if (contentType === EPISODE_CONTENT_KIND_MEDIA) {
    return `${PODCAST_DATA_KEY_PREFIX}${episodeId}/raw-media`;
  } else if (contentType === EPISODE_CONTENT_KIND_TEXT) {
    return `${PODCAST_DATA_KEY_PREFIX}${episodeId}transcription.txt`;
  }
};

const responseRedirect = async (client, bucket, key) => {
  try {
    const presigned = await getSignedUrl(
      client,
      new GetObjectCommand({
        Bucket: bucket,
        Key: key,
      }),
      {
        expiresIn: 24 * HOUR,
      }
    );
    return temporaryRedirectResponse(presigned);
  } catch (e) {
    throw new Error(`Failed to crate presigned URL ${contentType}: ${e}`);
  }
};

const checkMediaExists = async (client, bucket, key) => {
  try {
    await waitUntilObjectExists(
      {
        client,
        maxWaitTime: 15,
      },
      {
        Bucket: bucket,
        Key: key,
      }
    );
    return true;
  } catch (e) {
    console.error(`Failed to wait for object to exist: ${e}`);
    return false;
  }
};

File: lambda/python/play_podcast.py
import botocore.exceptions

from utils import (
    PODCAST_BUCKET, S3_PODCAST_PREFIX, HOUR_IN_SECONDS, get_s3_client,
    get_redirect_response, get_not_found_error_response
)


def lambda_handler(event, context):
    s3_bucket = PODCAST_BUCKET
    s3_key = _get_s3_artifact_key(event)
    s3_client = get_s3_client()
    artifact_exists = _wait_for_artifact_to_exist(s3_client, s3_bucket, s3_key)
    if not artifact_exists:
        return get_not_found_error_response(
            f'Episode media data not found, {s3_key}')
    url = _get_artifact_url(
        s3_client=s3_client, s3_bucket=s3_bucket, s3_key=s3_key)
    return get_redirect_response(url)


def _get_s3_artifact_key(event):
    artifact_name = 'raw-media'
    if event.get('queryStringParameters', {}).get('content') == 'text':
        artifact_name = 'transcription.txt'
    episode_id = event['pathParameters']['id']
    return f'{S3_PODCAST_PREFIX}{episode_id}/{artifact_name}'


def _get_artifact_url(s3_client, s3_bucket, s3_key):
    # TODO: Return presigned url for the 'get_object' operation using the
    #  s3_client.generate_presigned_url() method. This URL should be valid
    #  for 24 hours.
    return (
        f'https://s3.{s3_client.meta.region_name}.amazonaws.com/'
        f'{s3_bucket}/{s3_key}'
    )


def _wait_for_artifact_to_exist(s3_client, s3_bucket, s3_key):
    # TODO: Create an 'object_exists' waiter and wait at increments
    #  of three seconds for a maximum of six attempts. If the waiter
    #  fails/times out, the function should catch the WaiterError and return
    #  False.
    return True
